File: src/test/java/org/storck/simplex/util/CryptoUtilTest.java
package org.storck.simplex.util;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.security.KeyPair;
import java.security.PublicKey;

class CryptoUtilTest {

    @Test
    public void testGenerateKeypair() throws Exception {
        // When
        KeyPair result = CryptoUtil.generateKeyPair();

        // Then
        Assertions.assertNotNull(result, "Generated KeyPair should not be null");
        Assertions.assertNotNull(result.getPrivate(), "Generated private key should not be null");
        Assertions.assertNotNull(result.getPublic(), "Generated public key should not be null");
    }

    @Test
    public void testGenerateSignatureSuccessful() throws Exception {
        // Given
        KeyPair keyPair = CryptoUtil.generateKeyPair();
        byte[] input = new byte[] { 1, 2, 3, 4, 5 };

        // When
        byte[] result = CryptoUtil.generateSignature(keyPair.getPrivate(), input);

        // Then
        Assertions.assertNotNull(result, "Generated Signature should not be null");
    }

    @Test
    public void testVerifySignatureInvalid() throws Exception {
        // Given
        KeyPair keyPair = CryptoUtil.generateKeyPair();
        byte[] input = new byte[] { 1, 2, 3, 4, 5 };
        byte[] signature = CryptoUtil.generateSignature(keyPair.getPrivate(), input);
        // generate a different public key to get a wrong/bad verification
        PublicKey wrongPublicKey = CryptoUtil.generateKeyPair().getPublic();

        // When
        boolean result = CryptoUtil.verifySignature(wrongPublicKey, input, signature);

        // Then
        Assertions.assertFalse(result);
    }

    @Test
    public void testVerifySignatureValid() throws Exception {
        // Given
        KeyPair keyPair = CryptoUtil.generateKeyPair();
        byte[] input = new byte[] { 1, 2, 3, 4, 5 };
        byte[] signature = CryptoUtil.generateSignature(keyPair.getPrivate(), input);

        // When
        boolean result = CryptoUtil.verifySignature(keyPair.getPublic(), input, signature);

        // Then
        Assertions.assertTrue(result);
    }
}

File: simplex-core/src/test/kotlin/org/storck/simplex/util/CryptoUtilTest.kt
package org.storck.simplex.util

import io.kotest.matchers.shouldBe
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.nulls.shouldNotBeNull

class CryptoUtilTest : FunSpec({
    test("testGenerateKeypair") {
        val result = CryptoUtil.generateKeyPair()

        result.shouldNotBeNull()
        result.private.shouldNotBeNull()
        result.public.shouldNotBeNull()
    }

    test("testGenerateSignatureSuccessful") {
        val keyPair = CryptoUtil.generateKeyPair()
        val input = byteArrayOf(1, 2, 3, 4, 5)

        val result = CryptoUtil.generateSignature(keyPair.private, input)

        result.shouldNotBeNull()
    }

    test("testVerifySignatureInvalid") {
        val keyPair = CryptoUtil.generateKeyPair()
        val input = byteArrayOf(1, 2, 3, 4, 5)
        val signature = CryptoUtil.generateSignature(keyPair.private, input)
        val wrongPublicKey = CryptoUtil.generateKeyPair().public

        val result = CryptoUtil.verifySignature(wrongPublicKey, input, signature)
        result.shouldBe(false)
    }

    test("testVerifySignatureValid") {
        val keyPair = CryptoUtil.generateKeyPair()
        val input = byteArrayOf(1, 2, 3, 4, 5)
        val signature = CryptoUtil.generateSignature(keyPair.private, input)

        val result = CryptoUtil.verifySignature(keyPair.public, input, signature)
        result.shouldBe(true)
    }
})